Updates the subsystems controller and begins global var conversion.

PsiOmegaDelta opened the pull request:
Based on https://github.com/tgstation/tgstation/pull/25325.
Pros, as I understand it:
"Automatic" creation of a global vars list, without the need to run external scripts.
A well-defined initialization order of objects (as opposed to BYOND's internal which seems to do whatever it fancies).

<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->

PsiOmegaDelta committed:
Updates the subsystems controller and begins global var conversion.

Based on https://github.com/tgstation/tgstation/pull/25325.

PsiOmegaDelta commented on code/controllers/subsystems/garbage.dm:
```diff
@@ -307,9 +329,7 @@ var/datum/controller/subsystem/garbage_collector/SSgarbage
-	show_browser(usr, dat, "window=qdeletedlog")
-
-#define SearchVar(X) DoSearchVar(X, "Global: " + #X)
+	usr << browse(dat, "window=qdeletedlog")
```
This should be restored to show_browser

PsiOmegaDelta commented on code/controllers/subsystems/garbage.dm:
```diff
@@ -145,12 +143,36 @@ var/datum/controller/subsystem/garbage_collector/SSgarbage
+		highest_del_time = time
+	if (time > 10)
+		log_game("Error: [type]([refID]) took longer then 1 second to delete (took [time/10] seconds to delete)")
+		message_admins("Error: [type]([refID]) took longer then 1 second to delete (took [time/10] seconds to delete).")
```
Change to R_DEBUG output.

PsiOmegaDelta commented on code/controllers/master.dm:
```diff
@@ -107,7 +119,25 @@ var/CURRENT_TICKLIMIT = TICK_LIMIT_RUNNING
+				msg = "The [BadBoy.name] subsystem seems to be destabilizing the MC and will be offlined."
+				BadBoy.flags |= SS_NO_FIRE
+		if(msg)
+			to_chat(GLOB.admins, "<span class='boldannounce'>[msg]</span>")
```
Change to R_DEBUG output.

PsiOmegaDelta commented on code/controllers/globals.dm:
```diff
@@ -0,0 +1,59 @@
+	if(!force)
+		return QDEL_HINT_LETMELIVE
+
+	qdel_null(statclick)
```
Should probably uppercase ```qdel_null()``` if we're to follow proper macro naming convention.

PsiOmegaDelta commented on code/controllers/master.dm:
```diff
@@ -126,30 +156,30 @@ var/CURRENT_TICKLIMIT = TICK_LIMIT_RUNNING
 	sortTim(subsystems, /proc/cmp_subsystem_display)
 	// Set world options.
+#ifndef UNIT_TEST
 	world.sleep_offline = 1
```
I believe I found the reason for the stall.

PsiOmegaDelta commented:
By the time this one is done it's going to have thousands upon thousands of line alterations...

skull132 commented:
We racked and stacked 17k additions with 13k removals when we converted fully to SMC.

You can do it :P

PsiOmegaDelta commented:
This global var conversion alone might reach ±10k lines and I haven't even begun converting the processes yet.

PsiOmegaDelta committed:
Updates the subsystems controller and begins global var conversion.

Third commit. The intention is to make Github reviews doable by doing it on a per commit basis.

StarSmasher44 commented:
2 weeks later, how is this coming along? :)

PsiOmegaDelta commented:
Only have about 780 global vars left.

StarSmasher44 commented:
Sounds like you're nearly there, was this job as tedious as I imagined it to be? Hahaha

StarSmasher44 commented:
Gonna have to ask, how far off are we from an actually good processor?

PsiOmegaDelta commented:
The /tg/ subsystem is already in partial use. This is just an update.

StarSmasher44 commented:
Oh well then I was mistaken about the eventual purpose of this PR, when will the processor actually be used for more than just lighting/GC then?

PsiOmegaDelta commented:
This PR is just to upgrade the main processor and the GC subsystem to the latest version, which requires overhauling our entire global var handling.
Any additional work will be conducted in separate PRs after this.

PsiOmegaDelta committed:
Updates the subsystems controller and begins global var conversion.

Fourth commit. The intention is to make Github reviews doable by doing it on a per commit basis.

PsiOmegaDelta committed:
Updates the subsystems controller and begins global var conversion.

Fifth commit. The intention is to make Github reviews doable by doing it on a per commit basis.

PsiOmegaDelta committed:
Updates the subsystems controller and begins global var conversion.

Fifth commit. The intention is to make Github reviews doable by doing it on a per commit basis.

PsiOmegaDelta commented:
Updates the subsystem handling to tg's latest.
Initialization is now handled by the atoms subsystem.
Adds a unit test to ensure everything initializes correctly.
Adds a unit test to prevent pipes in bad directions (revealed by the tests above).

PsiOmegaDelta commented:
The global var port is only partially complete due to the complexities brought in by requiring the /tg/ initialization system, but this PR now in such a state that I can continue that work in a separate PR after this.

A test has been added to ensure that one cannot add new global vars using the classic method accidentally.

Tridale commented:
One minor issue I found through some quick testing on a local, the 'time to start' counter when in the lobby goes down really fast. Like roughly twice as fast, so the 180 timer is more like 90 seconds.

Round time and local (IC) time appear to count normally once round starts however.

Everything else through quick testing and messing around seems to be working fine, I'll continue to test more when I have more time (tomorrow)

PsiOmegaDelta commented:
Curious. Haven't really touched anything in the ticker except to add the game state calls.

Tridale reviewed (approved):
Looks good to my limited knowledge and some playtesting on a local shows nothing explodes nor seems noticeably more unstable. I'll continue messing around but unless I find something that can be pointed directly to this PR. I approve.
